### PR TITLE
Add typing to ``NodeNG.statement``

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

`statement()` is only defined twice. Once on `NodeNG` and once on `Module`. The one on `NodeNG` refers to `self.is_statement` which is only set to `True` on `Statement`. As a result `statement()` can either return `Module` or `Statement`.

Added because of comment by @cdce8p here:
https://github.com/PyCQA/pylint/pull/5163#pullrequestreview-786711662

I think I will not wait for this version to be added to `pylint` to update the typing there.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:262]: @cdce8p Pinging you again. Please consider unsent if you just haven't had time to look at this!

Comment (cdce8p) [reply astroid/nodes/node_ng.py:262]: > (Just passing a `future` argument to functions doesn't really work because we would then deprecate the future parameter afterwards.)

Isn't that how `SQLAlechemy` will do it, too? I would recommend to add it as argument to `statement()`. Passing it during initialization has some drawbacks which I'd like to avoid:
* Difficult to pass around to all nodes, could introduce issues with circular dependencies. Although we might be able to make it work in the end.
* What if we want to add a `future` argument to another function? We would break code that has already added the flag.

Comment (DanielNoord) [reply astroid/nodes/node_ng.py:262]: https://docs.sqlalchemy.org/en/14/core/future.html
https://github.com/sqlalchemy/sqlalchemy/blob/8656050241766688fee8363b5bb950d3945756c1/lib/sqlalchemy/engine/create.py#L627-L632

Sorry, I should have linked the documentation.
Seems to me that they pass the argument to two of their base classes as a "flag" and then methods/classes get changed depending on that flag. However, I have never used their code before, so it might not be true. 

I'm against passing it as an argument to `statement()` because then we would need to deprecate the `future` parameter again at some point. That seems like double-work?

File: astroid/nodes/scoped_nodes.py
Comment (Pierre-Sassoulas) [inline astroid/nodes/scoped_nodes.py:77]: I just realized that we also need to do https://github.com/PyCQA/pylint/issues/5065 for astroid.

**Review by Pierre-Sassoulas — APPROVED** [review]
👍

File: astroid/nodes/scoped_nodes.py
Comment (DanielNoord) [reply astroid/nodes/scoped_nodes.py:77]: I believe this is the only one. Shall I open an issue so we don't forget about this?

Comment (Pierre-Sassoulas) [reply astroid/nodes/scoped_nodes.py:77]: Might as well create an issue from the [unofficial todo list](https://github.com/PyCQA/pylint/issues/5065#issuecomment-935428793) (hidden in dozen of comment 😄) and add a "release astroid 2.8.5 for > 3.6.2 only" ?:

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py]: > Sorry, I should have linked the documentation. Seems to me that they pass the argument to two of their base classes as a "flag" and then methods/classes get changed depending on that flag. However, I have never used their code before, so it might not be true.

Yes, they pass it to the base functions. However, as I explained earlier, that doesn't really work for us. Should we add `future2`, `future3`, ... for any subsequent change to other functions? `SQLAlchemy` is different as they have just one big change that they what to do, not many small once.

> I'm against passing it as an argument to `statement()` because then we would need to deprecate the `future` parameter again at some point. That seems like double-work?

We need to deprecate it either way. If you don't like that, the only other alternatives I can think of would be to not add a breaking change or create a new method with a different name that replaces `statement()`. Not sure these are any better, though.

--
It would be a bit easier if we were changing function arguments. Then we could use something similar to what I've done for the removal of `None` as default values and use `inspect.signature` among other things. Without any `arguments` that doesn't work unfortunately.
https://github.com/PyCQA/astroid/blob/cdf8a2a9813128db5732a792434b89967c6fb701/astroid/decorators.py#L155-L209

Comment (DanielNoord) [reply astroid/nodes/node_ng.py]: > We need to deprecate it either way. If you don't like that, the only other alternatives I can think of would be to not add a breaking change or create a new method with a different name that replaces `statement()`. Not sure these are any better, though.

Would the `future` parameter be deprecated at the same time as we change the normal behaviour? I think I (wrongly) assumed that the flow would be:
1) Allow `future` parameter
2) Bump to 3.0 > change normal behaviour
3) Warn of deprecation for `future` parameter
4) Bump to 4.0 > disallow `future` parameter
 
Thinking about this you probably meant to do:
1) Allow `future` parameter
2) Bump to 3.0 > change normal behaviour and disallow `future` parameter

The second option seems fine to me!

Comment (cdce8p) [reply astroid/nodes/node_ng.py]: I believe `SQLAlchemy` uses the first flow, although they might not wait till the next major release to disallow it.
For us, the second option should work. It's just a matter of removing the `future=True` at that point.

Comment (DanielNoord) [reply astroid/nodes/node_ng.py]: I'll create a commit that implements this!

File: astroid/exceptions.py
Comment (DanielNoord) [inline astroid/exceptions.py:267]:
```suggestion
        target: The node for which the parent lookup failed.
        missing_element: The type of attribute/method that is missing because of the missing parent
```

**Comment by DanielNoord** [comment]
2 questions:
- Is the `pylint` message about `self = cast("nodes.Statement", self)` correct?
- How should I overload `nodes.Module.statement()`? I thought this was correct but both `pylint` and `mypy` complain..

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:269]: I believe we should add an additional overload before that.
```py
    @overload
     def statement(self) -> Union["nodes.Statement", "nodes.Module"]:
         ...
```
This should match those cases were `future` isn't passed.

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:273]: `NoReturn` should not be a return type here. As I understand it, it should only be used to annotate functions that **never** return. I.e. unconditional `raise` or `sys.exit`.

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:277]:
```suggestion
    ) -> Union["nodes.Statement", "nodes.Module"]:
```

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:287]:
```suggestion
            return cast("nodes.Statement", self)
```
The `pylint` warning is correct. In that case it should work by retuning the result of `cast` directly.
Ideally we wouldn't even need it. It might not be worth it as we risk circular dependencies, but would
`isinstance(self, nodes.Statement)` work? Might also be a breaking change if someone depends on manually assigning `is_statement`.

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:656]: Same here. Add an overload without `future`.

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:660]: Here it's correct.

**Review by cdce8p — COMMENTED** [review]
Just some initial thoughts. Need to come back to it later when I've more time.
What do you think about making `future` a keyword only argument?

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:269]: I think the `...` after `Literal[False]` does this? I played around with it and this does seem to work. VS Code picks up on it and `mypy` does as well I think.

The screenshot is from `nodes.Module`, but I think it should work similarly.
Schermafbeelding 2021-11-03 om 16 35 52

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:273]: I added it because I thought it would create problems with `nodes.Module.statement()` if its return type did not match either of the possibilities of `NodeNG`.
`Mypy` is turned off for `astroid`, but won't it complain when a class that inherits from `NodeNG` overwrites a method and returns something that is not included in the possibilities for `NodeNG`?

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py]: It already creates a circular dependency, that's why I opted to use `cast`.

I thought about the `pylint` warning and couldn't figure out why it would be correct. `typing.cast` returns the second argument without any changes, so that should be fine, right?
I opened https://github.com/PyCQA/pylint/pull/5254 (prematurely). If the warning is indeed correct I will change that PR.

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py:269]: Tested a few things. What do you think about this solution?
```py
    @overload
    def statement(
        self, future: Literal[None] = ...
    ) -> Union["nodes.Statement", "nodes.Module"]:
        ...

    @overload
    def statement(self, future: Literal[True]) -> "nodes.Statement":
        ...

    def statement(
        self, future: Optional[Literal[True]] = None
    ) -> Union["nodes.Statement", "nodes.Module"]:
```

And this for `Module`:
```py
    @overload
    def statement(self, future: Literal[None] = ...) -> "Module":
        ...

    @overload
    def statement(self, future: Literal[True]) -> NoReturn:
        ...

    def statement(self, future: Literal[True] | None = None) -> Union[NoReturn, "Module"]:
```

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py:273]: I see what you mean. Tested it with both `mypy` and `pylance` and both seem to except it. Might be that `NoReturn` is special cased somehow.

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py:269]: Note that I also changed the default to `None`.

Comment (DanielNoord) [reply astroid/nodes/node_ng.py:269]: Looks good to me! I'll create a commit.

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:273]: We learn something new every day :)

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:269]: Did you test these with `pylint`?
I'm getting the following error with pre-commit:

```shell
************* Module astroid.nodes.scoped_nodes
astroid/nodes/scoped_nodes.py:659:4: W0222: Signature differs from overridden 'statement' method (signature-differs)
```

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:663]:
```suggestion
        self, future: Optional[Literal[True]] = None
```

Comment (DanielNoord) [reply astroid/nodes/scoped_nodes.py]: Oops, I replaced the `|` from your suggestion. Should have payed more attention.

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py]: Turns out, you were right all along. Just tested it with strict mode and now `pylance` complains. Should be fixed by adding `NoReturn` just to the function signature in `NodeNG` (not as part of overloads).

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py]: I did not. It's not immediately obvious to me what's causing the issue. Need to investigate that a bit more.

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [reply astroid/nodes/scoped_nodes.py]: I should have used the correct syntax for `astroid`.

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py]: Added with
https://github.com/PyCQA/astroid/pull/1217/commits/6026560846c00ade3c6f2d2f89f71331300e10e4

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:276]:
```suggestion
        self, future: Literal[None, True] = None
```
Might be a bit easier to understand, especially with the first overload.

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:291]: Can we emit a `DeprecationWarning` if `future==None` and `AttributeError` is emitted?

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:659]:
```suggestion
    # pylint: disable-next=signature-differs
    # https://github.com/PyCQA/pylint/issues/5264
    @overload
    def statement(self, future: Literal[True]) -> NoReturn:
```

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:663]:
```suggestion
        self, future: Literal[None, True] = None
```

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:675]: Same here: `DeprecationWarning`

File: tests/unittest_builder.py
Comment (cdce8p) [inline tests/unittest_builder.py:618]: After the recent change `statement(False)` should never be called. Either `None` or `True`.

File: tests/unittest_nodes.py
Comment (cdce8p) [inline tests/unittest_nodes.py:633]:
```suggestion
```

File: tests/unittest_scoped_nodes.py
Comment (cdce8p) [inline tests/unittest_scoped_nodes.py:326]:
```suggestion
```

File: tests/unittest_scoped_nodes.py
Comment (cdce8p) [inline tests/unittest_scoped_nodes.py:826]:
```suggestion
```

**Review by cdce8p — COMMENTED** [review]
I think it would be useful if we emit `DeprecationWarnings` in case `future=True` isn't passed to `statement()`. For test cases, `unittest_decorators.py` could be helpful.

File: astroid/exceptions.py
Comment (DanielNoord) [reply astroid/exceptions.py:273]: Good idea!

Comment (DanielNoord) [reply astroid/exceptions.py]: This creates a new pylint error 😅 

Is this the correct way to do this?

File: tests/unittest_builder.py
Comment (cdce8p) [inline tests/unittest_builder.py:620]:
```suggestion
        with pytest.warns(DeprecationWarning) as records:
            module.statement()
            assert len(records) == 1
```
`warnings` is also the name of the stdlib module, so it might be better to use something different instead.

File: tests/unittest_nodes.py
Comment (cdce8p) [inline tests/unittest_nodes.py:633]:
```suggestion
        with (
            self.assertRaises(AttributeError),
            pytest.warns(DeprecationWarning) as records
        ):
            node.statement()
            assert len(records) == 1
```

File: astroid/nodes/node_ng.py
Comment (cdce8p) [inline astroid/nodes/node_ng.py:278]:
```suggestion
    def statement(
        self, *, future: Literal[None, True] = None
    ) -> Union["nodes.Statement", "nodes.Module", NoReturn]:
```

I should have brought that up earlier but I somehow missed it. What do you think about making `future` a keyword only arg? That will make it explicit, prevent issue if we decide to add additional arguments, and make it easier to deprecate / remove.

**Review by cdce8p — COMMENTED** [review]
I've seen that `pytest` will emit `DeprecationWarnings` for `astroid` itself. Could you add `future=True` to existing statement in `astroid`, too?
https://github.com/PyCQA/astroid/runs/4126645345?check_suite_focus=true#step:6:81

File: astroid/exceptions.py
Comment (cdce8p) [reply astroid/exceptions.py]: We seem to be hitting quite a streak with false-positives 😂
This is valid, although at the limit of what one should reasonable do.
```py
class Parent:
    def __init__(self):
        print("Parent")

class Child(Parent):
    def __init__(self):
        print("Child")
        super().__init__()

class GrandChild(Child):
    def __init__(self):
        print("Grandchild")
        super(Child, self).__init__()  # will skip 'Child.__init__` and call 'Parent.__init__' instead

GrandChild()
```

**Links and related issues**
* https://realpython.com/python-super/#a-super-deep-dive
* https://github.com/PyCQA/pylint/issues/2903 (We might what to consider reopening this one)
* https://github.com/PyCQA/pylint/issues/4922

Comment (cdce8p) [reply astroid/exceptions.py]: If you don't want to add `pylint: disable`, I think the alternative would be to add `message: Optional[str] = None` to `ParentMissingError`. And pass conditionally pass either `message` or the new constructed string to `super().__init__(...)`.

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:295]: I added something about `AttributeError` but think it might be helpful to fully explain what `statement()` will now do. Saves the maintainer/contributor changing the code some time!

**Comment by DanielNoord** [comment]
> I've seen that `pytest` will emit `DeprecationWarnings` for `astroid` itself. Could you add `future=True` to existing statement in `astroid`, too?
> https://github.com/PyCQA/astroid/runs/4126645345?check_suite_focus=true#step:6:81

This started to fail some of the tests. I wanted to do this in a separate PR because I wondered if some of the functions/methods calling `statement()` are considered `public` and should thus keep raising `AttributeError` or returning `Module`.

File: astroid/exceptions.py
Comment (DanielNoord) [reply astroid/exceptions.py]: I would be in favour of opening PyCQA/pylint#2903. I also googled online (and found the example you use) and some others. Seems that this is considered valid pythonic code.

Shall I reopen and add a `disable` + reference to the issue?

Comment (DanielNoord) [reply astroid/exceptions.py]: I would be in favour of opening PyCQA/pylint#2903. I also googled online (and found the example you use) and some others. Seems that this is considered valid pythonic code.

Shall I reopen and add a `disable` + reference to the issue?

File: tests/unittest_nodes.py
Comment (DanielNoord) [reply tests/unittest_nodes.py]: `black` failed on these combined `with` statements. It seems like it can't parse those statements together. I reverted in the latest commit but kept using `records`.

File: astroid/exceptions.py
Comment (cdce8p) [reply astroid/exceptions.py]: Sounds good to me!

File: astroid/nodes/node_ng.py
Comment (cdce8p) [reply astroid/nodes/node_ng.py]: 👌🏻

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py:268]: Ah forgot about that!

File: tests/unittest_nodes.py
Comment (cdce8p) [reply tests/unittest_nodes.py]: I realized that it also requires at least Python 3.9. So even if black would work, we couldn't use it.
For reference: https://github.com/psf/black/issues/1948

This would work, but looks rather ugly with black formatting.
```py
        with self.assertRaises(AttributeError), pytest.warns(
            DeprecationWarning
        ) as records:
            node.statement()
            assert len(records) == 1
```
Let's leave it like it is then.

File: astroid/nodes/node_ng.py
Comment (DanielNoord) [reply astroid/nodes/node_ng.py]: Done now!

File: astroid/exceptions.py
Comment (DanielNoord) [reply astroid/exceptions.py]: Done!

Comment (cdce8p) [reply astroid/exceptions.py]: Thanks!

File: astroid/exceptions.py
Comment (cdce8p) [inline astroid/exceptions.py:287]:
```suggestion
        # pylint: disable-next=bad-super-call
        # https://github.com/PyCQA/pylint/issues/2903
        # https://github.com/PyCQA/astroid/pull/1217#discussion_r744149027
```
That's usually how I do it.

**Review by cdce8p — APPROVED** [review]
Overall, LGTM! Thanks @DanielNoord 🚀
A bit unfortunate, that we came across multiple false-positives in the process.

@Pierre-Sassoulas Do you want to take a final look at it? Maybe with focus on the Exception and deprecation message?

File: astroid/exceptions.py
Comment (DanielNoord) [reply astroid/exceptions.py:287]: Oh wow, does this work?
That's good to know for https://github.com/PyCQA/pylint/issues/5258

Comment (DanielNoord) [reply astroid/exceptions.py:287]: Oh wow, does this work?
That's good to know for https://github.com/PyCQA/pylint/issues/5258

**Comment by DanielNoord** [comment]
Thanks for all of your help here @cdce8p!

Indeed a bit unfortunate, but I learned a lot from this PR and it should help with some of the future PR's and issues we might come across.

File: astroid/exceptions.py
Comment (Pierre-Sassoulas) [inline astroid/exceptions.py:286]: I think the spirit of the check is to signal the problem to people that don't really know how to use super(). Here it's a false positive because calling the father's function is what you really want to do, but maybe in most case and especially for beginner  the check is right. If we remove this "false positive" the check will only have really obvious cases where the class called is not even an ancestor ?

**Review by Pierre-Sassoulas — APPROVED** [review]
I only have a small remark regarding the "bad super call" but I don't see anything wrong with the MR otherwise. To be fair I did not find the time to read the whole conversation and this seem really polished :) Look like you took care of a lot of problems, thanks a lot for handling this !

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [inline astroid/nodes/scoped_nodes.py:660]:
```suggestion
```
It seems like your luck didn't outran us yet. This actual now triggers a `useless-suppression` message. Guess we found a `false-negative` by adding `*` 😅 https://github.com/PyCQA/pylint/issues/5270

File: astroid/exceptions.py
Comment (cdce8p) [reply astroid/exceptions.py:286]: 🤔 Not sure about that. What it comes done to is if we want to emit errors for valid Python code. Even if it's probably not what was intended, there are still situations where it's useful and has it's place, especially with multi-inheritance.

The check also emits errors in cases that could cause recursion loops:
```py
class SuperWithType(object):
    """type(self) may lead to recursion loop in derived classes"""
    def __init__(self):
        super(type(self), self).__init__()  # [bad-super-call]

class SuperWithSelfClass(object):
    """self.__class__ may lead to recursion loop in derived classes"""
    def __init__(self):
        super(self.__class__, self).__init__()  # [bad-super-call]
```
Any modifications should also make sure to still emit errors if the class is not part of the MRO.
```py

class Getattr(object):
    """ crash """
    name = NewAaaa

class CrashSuper(object):
    """ test a crash with this checker """
    def __init__(self):
        super(Getattr.name, self).__init__()  # [bad-super-call]
```

Test cases taken from the `pylint` testsuit: [super_checks.py](https://github.com/PyCQA/pylint/blob/0a1ebd488fcdf7bf306615aba29e401ce49e3e10/tests/functional/s/super/super_checks.py)

Comment (Pierre-Sassoulas) [reply astroid/exceptions.py:286]: OK, I was trying to search for the initial issue where this was implemented for the rational but I can't find it. It makes sense to not warn for valid code, and the check has a lot of valid test cases, so let's fix this false positive.

File: astroid/nodes/scoped_nodes.py
Comment (DanielNoord) [reply astroid/nodes/scoped_nodes.py:660]: Is `useless-suppression` not enabled for `astroid`? CI is passing.

We might want to enable that in another PR.

Comment (cdce8p) [reply astroid/nodes/scoped_nodes.py]: It's enabled, but not a failure condition! I don't think we should change that.

Comment (DanielNoord) [reply astroid/nodes/scoped_nodes.py]: Sure? Seems like something that will be easily missed in review

Comment (Pierre-Sassoulas) [reply astroid/nodes/scoped_nodes.py]: I also think we should make it a failure condition so it's handled in the proper MR each time.

**Review by cdce8p — APPROVED** [review]
Let's (squash) merge it!

File: astroid/nodes/scoped_nodes.py
Comment (cdce8p) [reply astroid/nodes/scoped_nodes.py]: > We might want to enable that in another PR.

It would be as easy as adding `--fail-on=useless-suppression` to the `pylint` command.

The issue is that some errors might be Python version specific. It could create a situation where only one, `pre-commit` or `CI`, will succeed when tested with different versions. The `py-version` option should help, but I don't know if it's enough.

Additionally, this would prevent us from having the latest `pylint` version installed in a local env as this is likely to have fixed some false-positives. -> Which would then trigger `useless-suppression`.

Comment (DanielNoord) [reply astroid/nodes/scoped_nodes.py]: That last point will indeed be difficult to avoid. I'll try and think of something to fix that, but in the meantime let's let this rest!

Comment (cdce8p) [reply astroid/nodes/scoped_nodes.py]: Do you maybe want to track this as issue in `pylint`?

> in the meantime let's let this rest!

I completely agree!

Comment (DanielNoord) [reply astroid/nodes/scoped_nodes.py]: Shouldn't this be tracked in `astroid`? `pylint` has its own issues with `useless-suppression` (https://github.com/PyCQA/pylint/issues/5040) 😄

Comment (cdce8p) [reply astroid/nodes/scoped_nodes.py]: 😂 Though if we enable it, we'll likely want to do it for `pylint` too. Tbh I don't really care about it due to the issues mentioned earlier.

**Pull request merged**
